Extract autos after LST binning

mwilensky768 opened the pull request:
This adds a task script to the H4C analysis that extracts autocorrelations into a big waterfall mainly for the purposes of error propagation.

I am unsure as to whether I should have made a new toml, or tacked this on to the rerun branch.

Commit:
Add extract autos do script

Commit:
Add EXTRACT_AUTOS to workflow script

Commit:
add chunk_size option to auto extractor

r-pascua commented on pipelines/h4c/post_processing/v7_interleave/post_lstbin_frate/task_scripts/do_EXTRACT_AUTOS.sh:
```diff
@@ -0,0 +1,20 @@
+    
+    echo "python extract_autos_post_lstbin.py ${sd} ${label} --flist ${flist}"
+    
+    python extract_autos.py ${flist}
```
The script you're "calling" in the `echo` does not match the script you're actually calling, and the arguments don't match.

r-pascua commented on pipelines/h4c/post_processing/v7_interleave/post_lstbin_frate/task_scripts/do_EXTRACT_AUTOS.sh:
```diff
@@ -0,0 +1,20 @@
+
+for sd in ${sumdiff[@]}
+do
+    flist=$(ls *${sd}*chunked.uvh5)
```
Have you checked that this gives an empty list when the particular file type (either 'sum' or 'diff') is missing? Additionally, have you tested that the auto extraction script doesn't break when passed an empty list?

r-pascua reviewed (commented):
I think this is mostly OK, but I think the do script might need a bit more development. I'd say let's hold off on this until the script is settled and lives somewhere. After that, we should come back to this.

That said, I think just adding another item to the toml and a corresponding do script is the right way to do it :)

Thanks for your work!

Commit:
port auto extraction script wholesale from pspec branch

Commit:
rewrite script to avoid convoluted try/except logic

Commit:
make call match echo

Commit:
Add extra args in call and actually make call match

Commit:
Fix typos

Commit:
Add include_diffs option

Commit:
add include_diffs handling and ls error handling

mwilensky768 commented:
@r-pascua I moved the extract_autos_post_lstbin script to this repo for now. 

I rewrote it so that it does a metadata-only read on the passed list and makes a list of all the files that have autos. This should be fast since it's metadata-only, and gets rid of the fragile error handling in the original version. I then just read in all the autos at once with select on read and optional `axis` kwarg. Made `clobber` optional too. I think this addresses the concerns from your review on the pspec branch. 

I also made doing diffs optional at the do script level, and then did some error handling on the `ls` call to make sure an empty list doesn't get passed to the python script. The basic logic is that `ls` returns `0` if there are matching files and something greater than `0` for any other error (recall that exiting with `0` has a confusing truth value in bash), so if `ls` succeeds, it assigns the list of files to `flist`, otherwise it exits with some verbiage since something went badly wrong.

r-pascua commented on pipelines/h4c/post_processing/v7_interleave/post_lstbin_frate/task_scripts/do_EXTRACT_AUTOS.sh:
```diff
@@ -0,0 +1,36 @@
+    # Use the return value of ls to see if it runs OK (don't want to pass an empty list for example)
+    # Redirect the output to die in /dev/null/ so it doesn't crowd the log
+    # Exit if there is an error.
+    if ls *${sd}.${label}.*chunked.uvh5 1> /dev/null 
+    then
+        flist=$(ls *${sd}.${label}.*chunked.uvh5)
+    else
+        echo "Error when calling ls on ${sd} files with label ${label}. Probably could not find files. Check error log. Exiting."
+        exit 1
```
I'm a little confused about this, in the sense that I thought the "chunked" files would be the ones that have all of the times for a chunk of baselines and one spectral window, and so we wouldn't really want to look for those files&mdash;I would think that instead we would want to look for whatever the files look like fresh out of the LST binner.

Also, have you tested the magic on line 25? It's not obvious to me that it'll do the right thing (but I'm not a bash expert, so maybe it is obvious to someone with more experience).

mwilensky768 replied:
Files with all the times for a small number of baselines are called "waterfall" files, and these are currently not output in the unfiltered state. I need the unfiltered autos, because the FRF'd autos are not the mathematically correct item to use for error propagation. You have to do a slightly different, but related operation to the unfiltered autos to get the right covariances. 

The files fresh out of the LST-binner have 2 integrations each, but lots of baselines. The chunked files have fewer baselines but 20 integrations each. There is no processing being done between these as far as I know. 

I think it is less files to read in using the chunked files, so these are the files we should use for this step. 

Yeah, line 25 works. The exact meaning of everything is in some commented lines above the magic.

r-pascua replied:
Great, thanks for the explanation! My main concern is that I don't have a sense of whether we ought to expect the chunked files to have the autos in them, but I figure that's why you go through the trouble of checking the files for autos.

r-pascua commented on pipelines/h4c/post_processing/v7_interleave/post_lstbin_frate/h4c_idr1_filtering_allbands_for_lstbin_pspec_nocleanup_frate_filter.toml:
```diff
@@ -70,6 +71,13 @@ stride_length="${Options:chunk_size}"
+[EXTRACT_AUTOS]
+prereqs = "PRE_CHUNK"
+prereq_chunk_size="all"
+chunk_size="all"
```
I'm not sure this is the right way to do it. I think it's probably better to take inspiration from the FRF section, where we set `chunk_size=1` and set the `stride_length` to the number of files in a night. I think invoking it as-is would cause makeflow to generate one `EXTRACT_AUTOS_blah.sh` file for every file. If I'm understanding the intended workflow correctly, I'd think that we only want to run the script once.

mwilensky768 replied:
I think you understand the workflow correctly. I just didn't find the documentation for these "keywords" that clear so I guessed :)

Will change it. By the way, it seems like that number basically has to be hardcoded (like the long list of options at the top) unless there's a way of exporting an environment variable into the toml.

mwilensky768 replied:
(See new toml)

r-pascua replied:
Hmmm. I'm not sure whether it's possible to export environment variables into the toml. Tbh, I'm very fuzzy on the details of how the chunking stuff even works with makeflow, and just figured that we could mimic what the FRF code does since it operates on an entire night's worth of data. Maybe @jsdillon can provide further insights?

r-pascua commented on pipelines/h4c/post_processing/v7_interleave/post_lstbin_frate/task_scripts/do_EXTRACT_AUTOS.sh:
```diff
@@ -0,0 +1,36 @@
+# This script extracts the autocorrelations into a single waterfall file
+label=${1}
+include_diffs=${2}
+
```
It's possible that we'll need to manually move into the directory with the files. Figured I should flag it here as a reminder in case things break in the future and it's not obvious why.

r-pascua commented on scripts/extract_autos_post_lstbin.py:
```diff
@@ -0,0 +1,67 @@
+        Check that a file being read is part of the intended group. Error if not,
+        since something has been passed incorrectly.
+        """
+        if not args.sumdiff in file:
+            raise ValueError(f"Supposedly processing {args.sumdiff} files but "
+                             f"{args.sumdiff} not in the filename.")
+        if not args.label in file:
+            raise ValueError(f"Supposedly processing {args.label} files but "
```
You can instead write `if args.sumdiff not in file` (and do a similar thing for the other check) to make this a little more readable.

mwilensky768 replied:
Done.

r-pascua commented on scripts/extract_autos_post_lstbin.py:
```diff
@@ -0,0 +1,67 @@
+    files_with_autos = []
+    for file in args.flist:
+        check_for_sumdiff_label(file)
+        test_uvd = UVData()
+        test_uvd.read(file, read_data=False)
```
This can be combined into a single line via `test_uvd = UVData.from_file(file, read_data=False)`

mwilensky768 replied:
Done, thanks.

r-pascua commented on scripts/extract_autos_post_lstbin.py:
```diff
@@ -0,0 +1,67 @@
+    
+    assert len(files_with_autos) > 0, "No files with autos found. Check inputs."
+
+    auto_uvd = UVData() 
+    auto_uvd.read(files_with_autos, ant_str="auto", axis=args.axis)
```
Same thing here, you can do `auto_uvd = UVData.from_file(...)`

mwilensky768 replied:
Done

r-pascua commented on scripts/extract_autos_post_lstbin.py:
```diff
@@ -0,0 +1,67 @@
+    
+    # Just formatting this string so it doesn't make a long line
+    prefix = "zen.LST.0.00000"
+    suffix = "foreground_filled.xtalk_filtered.chunked.waterfall.autos.uvh5"
```
Have these files actually been filtered? I thought the filtering scripts tended to skip over the autos.

mwilensky768 replied:
No, they have not been filtered in this script, so I did not put an "frf" tag, but I did put a "waterfall" tag, because they now have all the times in one file.

r-pascua replied:
I think I wasn't clear enough with my comment: I was addressing the fact that "xtalk_filtered" is in `suffix`, so it could be a somewhat misleading file name.

mwilensky768 replied:
\sigh

Right, thanks, that makes sense. Do you know offhand if the xtalk filter (different than the frf step...) skips over the autos? 

If so, I will remove `xtalk_filtered` from the output filename as you suggest. 

If not, I will need to change the script to operate on a different set of files entirely (e.g. different wildcard in the bash script).

In either case, I think I should remove `chunked` from the filename, since `waterfall` should just replace that tag.

mwilensky768 replied:
After much digging through the pipeline scripts, I've learned that the xtalk_filter step indeed skips the autos. I will update the tag accordingly and merge when I get a chance.

r-pascua reviewed (commented):
This is looking nice, thanks @mwilensky768! I just have a few comments that I wanted to point out. Like Josh said in the pspec call, we'll most likely just need to try running this a few times to work out all of the bugs, but this is looking pretty close to doing the right thing.

Commit:
Change chunk_size, stride_length

Commit:
Make script more readable

r-pascua commented:
This is looking really nice, thanks @mwilensky768! I'd say merge this whenever you're happy with it. I'm sure we'll need to do some rounds of debugging, but that can be done without needing to go through a formal PR procedure (this is just a little handy for working out some kinks ahead of time and keeping track of the relevant discussions).

Commit:
Remove bad tags

Commit:
sort list for axis=blt
